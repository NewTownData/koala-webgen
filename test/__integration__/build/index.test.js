/**
 * This file must be placed in `__integration__` folder as the test cannot be run from watcher.
 * Creation of files cases watcher to crash jest.
 */
const path = require('path');
const fs = require('fs');
const { TextDecoder } = require('util');
const build = require('../../../src/build');
const now = require('../../../src/context/now');

jest.mock('../../../src/context/now');

const tempRoot = path.join(__dirname, '.temp_build');
const destination = path.join(tempRoot, 'dist');

const websiteRoot = path.join(__dirname, '..', '..', '..', 'website');

const TestDate = new Date(Date.UTC(2020, 11, 1, 14, 10, 20));

function utf8(buffer) {
  const decoder = new TextDecoder('utf-8');
  return decoder.decode(buffer);
}

describe('build', () => {
  beforeEach(() => {
    fs.mkdirSync(tempRoot, { recursive: true });

    now.mockClear();
    now.mockImplementation(() => TestDate);
  });

  afterEach(() => {
    fs.rmSync(destination, {
      recursive: true,
    });
    fs.rmSync(tempRoot, { recursive: true });
  });

  it('works', () => {
    expect(fs.existsSync(destination)).toBe(false);

    build(websiteRoot, destination);

    expect(fs.existsSync(destination)).toBe(true);
    expect(fs.readdirSync(destination)).toEqual([
      'about.html',
      'archive',
      'favicon.ico',
      'feed-style.xsl',
      'feed.rss.xml',
      'hello-world.html',
      'images',
      'index.html',
      'post1.html',
      'post2.html',
      'post3.html',
      'post4.html',
      'post5.html',
      'posts',
      'sitemap.txt',
      'style.css',
      'tags',
      'test',
    ]);
    expect(fs.readdirSync(path.join(destination, 'images'))).toEqual([
      'logo-128.png',
      'logo-512.png',
    ]);
    expect(fs.readdirSync(path.join(destination, 'archive'))).toEqual([
      '2019-01',
      '2019-03',
      '2020-06',
    ]);
    expect(
      fs.readdirSync(path.join(destination, 'archive', '2019-03')),
    ).toEqual(['page-1.html']);
    expect(fs.readdirSync(path.join(destination, 'tags'))).toEqual([
      'hello',
      'intro',
      'number',
      'test',
    ]);
    expect(fs.readdirSync(path.join(destination, 'tags', 'test'))).toEqual([
      'page-1.html',
    ]);
    expect(fs.readdirSync(path.join(destination, 'posts'))).toEqual([
      'page-2.html',
    ]);
    expect(fs.readdirSync(path.join(destination, 'test'))).toEqual([
      'test.html',
    ]);

    expect(
      utf8(fs.readFileSync(path.join(destination, 'feed.rss.xml'))),
    ).toMatchSnapshot();

    expect(
      utf8(fs.readFileSync(path.join(destination, 'sitemap.txt'))),
    ).toMatchSnapshot();
  });
});
