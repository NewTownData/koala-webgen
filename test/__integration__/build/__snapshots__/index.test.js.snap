// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build works 1`] = `
"<?xml version="1.0" encoding="UTF-8" ?>
<?xml-stylesheet href="/feed-style.xsl" type="text/xsl"?>
<rss version="2.0">
  <channel>
    <title>Koala - Static Website Generator</title>
    <description>A simple static website generator</description>
    <link>https://www.example.com/</link>
    <lastBuildDate>Tue, 01 Dec 2020 14:10:20 GMT</lastBuildDate>
    <pubDate>Tue, 01 Dec 2020 14:10:20 GMT</pubDate>
    <item>
      <title>Hello World</title>
      <description>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet, diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan. Aenean eu lacus quam. Morbi consequat ante quis eros iaculis tempus. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Nulla sit amet urna vitae elit iaculis blandit. Mauris dignissim consectetur augue vel varius.</description>
      <link>https://www.example.com/hello-world.html</link>
      <guid>https://www.example.com/hello-world.html</guid>
      <pubDate>Sat, 20 Jun 2020 10:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Test</title>
      <description>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet, diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan.</description>
      <link>https://www.example.com/test/test.html</link>
      <guid>https://www.example.com/test/test.html</guid>
      <pubDate>Wed, 10 Jun 2020 10:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 5</title>
      <description>Content 5</description>
      <link>https://www.example.com/post5.html</link>
      <guid>https://www.example.com/post5.html</guid>
      <pubDate>Sun, 03 Mar 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 4</title>
      <description>Content 4</description>
      <link>https://www.example.com/post4.html</link>
      <guid>https://www.example.com/post4.html</guid>
      <pubDate>Sat, 02 Mar 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 3</title>
      <description>Content 3</description>
      <link>https://www.example.com/post3.html</link>
      <guid>https://www.example.com/post3.html</guid>
      <pubDate>Fri, 01 Mar 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 2</title>
      <description>Content 2</description>
      <link>https://www.example.com/post2.html</link>
      <guid>https://www.example.com/post2.html</guid>
      <pubDate>Wed, 02 Jan 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 1</title>
      <description>Content 1</description>
      <link>https://www.example.com/post1.html</link>
      <guid>https://www.example.com/post1.html</guid>
      <pubDate>Tue, 01 Jan 2019 16:45:00 GMT</pubDate>
    </item>
  </channel>
</rss>"
`;

exports[`build works 2`] = `
"https://www.example.com/
https://www.example.com/about.html
https://www.example.com/hello-world.html
https://www.example.com/post1.html
https://www.example.com/post2.html
https://www.example.com/post3.html
https://www.example.com/post4.html
https://www.example.com/post5.html
https://www.example.com/test/test.html"
`;
